src/V3SplitVar.cpp

```diff
@@ -0,0 +1,1258 @@
+    explicit SplitPackedVarVisitor(AstNetlist* nodep)
+        : m_netp(nodep)
+        , m_modp(NULL)
+        , m_numSplit(0) {
+        iterate(nodep);
+    }
+    // When reusing the information from SplitUnpackedVarVisitor
+    SplitPackedVarVisitor(AstNetlist* nodep, RefsMap& refs)
+        : m_netp(nodep)
+        , m_modp(NULL)
+        , m_numSplit(0) {
+        for (RefsMap::iterator it = refs.begin(), it_end = refs.end(); it != it_end; ++it) {
+#if 1
+            m_modp = it->first;
+            it->second.visit(this);
+            split();
+            m_modp = NULL;
+#else
+            iterate(it->first);
+#endif
+        }
+    }
+    ~SplitPackedVarVisitor() {
+        UASSERT(m_refs.empty(), "Forgot to call split()");
+        V3Stats::addStat("SplitVar, Split packed variables", m_numSplit);
+    }
+
+    // Check if the passed variable can be split.
+    // Even if this function returns true, the variable may not be split
+    // when the access to the variable cannot be determined statically.
+    static const char* cannotSplitReason(const AstVar* nodep, bool checkUnpacked) {
+        const char* reason = NULL;
+        if (AstBasicDType* const basicp = nodep->dtypep()->basicp()) {
+            const std::pair<uint32_t, uint32_t> dim = nodep->dtypep()->dimensions(false);
+            // Unpacked array will be split in SplitUnpackedVarVisitor() beforehand
+            if (!((!checkUnpacked || dim.second == 0) && nodep->dtypep()->widthMin() > 1))
+                reason = "its bitwidth is 1";
+            if (!reason && !basicp->isBitLogic())  // Floating point and string are not supported
+                reason = "it is not an aggregate type of bit nor logic";
+            if (!reason) reason = cannotSplitVarCommonReason(nodep);
+        } else {
+            reason = "its type is unknown";
+        }
+        if (reason) UINFO(5, "Check " << nodep->prettyNameQ()
+                          << " cannot split because" << reason << endl);
+        return reason;
+    }
+    VL_DEBUG_FUNC;  // Declare debug()
+};
+
+static const char* cannotSplitPackedVarReason(const AstVar* varp) {
+    return SplitPackedVarVisitor::cannotSplitReason(varp, true);
+}
+
+//######################################################################
+// Split class functions
+
+void V3SplitVar::splitVariable(AstNetlist* nodep) {
+    UINFO(2, __FUNCTION__ << ": " << endl);
+    RefsMap refs;
+    {
+        SplitUnpackedVarVisitor visitor(nodep);
+        refs = visitor.getPackedVarRefs();
+    }
+    V3Global::dumpCheckGlobalTree("split_var", 0, v3Global.opt.dumpTreeLevel(__FILE__) >= 9);
+    { SplitPackedVarVisitor visitor(nodep, refs); }
+    V3Global::dumpCheckGlobalTree("split_var", 0, v3Global.opt.dumpTreeLevel(__FILE__) >= 9);
+}
+
+bool V3SplitVar::canSplitVar(const AstVar* varp) {
+    return !SplitUnpackedVarVisitor::cannotSplitReason(varp)
+           || !SplitPackedVarVisitor::cannotSplitReason(varp, false);
```

Comment:
Should the || be &&? If not please add comment to code as to why.

Reply:
`||` is correct.
If either SplitUnpackedVarVisitor or SplitPackedVarVisitor can handle it, that's fine.
I wrote in the comment.